File: mcdowell/python/ch9-10.py
cache = {}
G = {
    'counter': 0
}

def createStack(boxes, bottom):

    if bottom in cache:
        return cache[bottom]

    G['counter'] += 1

    maxHeight = 0
    maxStack = []

    for i in range(len(boxes)):
        if boxes[i] > bottom:
            newStack = createStack(boxes, boxes[i])
            newHeight = sum(newStack)

            if newHeight > maxHeight:
                maxHeight = newHeight
                maxStack = newStack


    maxStack.insert(0, bottom)
    cache[bottom] = maxStack
    return maxStack




print createStack([1,2,3,4,5], 1)
print G['counter']

File: mcdowell/js/ch9-10.js
var cache = {};


function createStack(boxes, bottom) {
    if (typeof cache[bottom] != 'undefined') {
        return cache[bottom];
    }

    var maxHeight = 0;
    var maxStack = [];

    for ( var i = 0; i < boxes.length; i ++ ) {
        if (boxes[i] > bottom) {
            var newStack = createStack(boxes, boxes[i]);
            var newHeight = newStack.reduce(function(pv, cv) { return pv + cv; }, 0);
            if (newHeight > maxHeight) {
                maxHeight = newHeight;
                maxStack = newStack;
            }
        }
    }

    maxStack.unshift(bottom);
    cache[bottom] = maxStack;
    return maxStack;
}


console.log(createStack([1,2,3,4,5,6], 1))